python/ray/includes/object_ref.pxi

```diff
@@ -134,6 +134,45 @@ cdef class ObjectRef(BaseID):
     def __await__(self):
         return self.as_future(_internal=True).__await__()
 
+    def __reduce__(self):
+        """Note: this code path is hit by pickle, but not ray.cloudpickle."""
+        raise RuntimeError(
+            "Object references cannot be pickled except via `ray.put()` or "
+            "passing as arguments to Ray tasks and actors. Use "
+            "`object_ref.export()` to export a reference for out-of-band use.")
+
+    def export(self, pin: bool = False) -> bytes:
```

Comment:
@stephanie-wang an alternative design might be to just allow serialization, but never pin the object. We say it's the user's responsibility to hold the reference (e.g., by adding to a global list in the current process).